presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -344,31 +344,12 @@ private void processNewSources()
         try {
             processNewSources();
 
-            // special handling for drivers with a single operator
-            if (activeOperators.size() == 1) {
-                if (driverContext.isDone()) {
-                    return NOT_BLOCKED;
-                }
-
-                // check if operator is blocked
-                Operator current = activeOperators.get(0);
-                Optional<ListenableFuture<?>> blocked = getBlockedFuture(current);
-                if (blocked.isPresent()) {
-                    current.getOperatorContext().recordBlocked(blocked.get());
-                    return blocked.get();
-                }
-
-                // there is only one operator so just finish it
-                current.getOperatorContext().startIntervalTimer();
-                current.finish();
-                current.getOperatorContext().recordFinish();
-                return NOT_BLOCKED;
-            }
-
+            // If there is only one operator, finish it
             // Some operators (LookupJoinOperator and HashBuildOperator) are broken and requires finish to be called continuously
-            // TODO remove when these operators are fixed
+            // TODO remove the second part of the if statement, when these operators are fixed
             // Note: finish should not be called on the natural source of the pipeline as this could cause the task to finish early
-            if (!activeOperators.isEmpty() && activeOperators.size() != allOperators.size()) {
+            if (activeOperators.size() == 1 ||
```

Comment:
You have `activeOperators.size() == 1` here. Is it necessary to have driver special handling for single operator?

In the for loop below, `finish` method of the first operator in general don't get called. Instead, its `isFinish` method will return true by itself at some point.

I don't see the inherent need for special handling for driver with only one operator. Every time I read this code, this special handling incurs a non-trivial cognitive overhead. That's why I'm proposing removal of it. If we still have `activeOperators.size() == 1` here, nothing really changes.

Reply:
We don't call finish on the first operator. I understand the question is: if there is (and always has been) exactly one operator, do we call finish on it?
This is definitely a special-case. @dain changes is an improvement (much less code to cover the special case). However, as i understand @haozhun's point, there is no need for this special case. If indeed so, why not just add a check in the constructor, rejecting operator list if shorter than 2?